### ISSUE TITLE
bluez-generated: Error: DbusError(D-Bus error: Method "WriteValue" with signature "aya{sv}" on interface "org.bluez.GattCharacteristic1" doesn't exist

### ISSUE THREAD

**Issue by ruediger** [alsuren/mijia-homie#91]
Hello,

I'm trying to use bluez-generated to talk to a meat thermometer. Thank you for writing buez-generated and I also borrowed your bluetooth.rs code. However I seem to have trouble writing to a GattCharacteristic. After finding and connecting to the device, I'm trying to call WriteValue. But I get an error

```
Error: DbusError(D-Bus error: Method "WriteValue" with signature "aya{sv}" on interface "org.bluez.GattCharacteristic1" doesn't exist
 (org.freedesktop.DBus.Error.UnknownObject))
```

I used QDBusViewer to look at the interface and it does seem to have that method:

```
<method name="WriteValue"><arg name="value" type="ay" direction="in"/>
<arg name="options" type="a{sv}" direction="in"/>
</method>
```

Any idea or suggestion on what I could be doing wrong? (I suspect that this is not really an issue in bluez-generated. But I thought I start here).

My code: https://github.com/ruediger/cloudbbq2-rs/blob/main/src/main.rs#L98

Thanks in advance.

(btw. would you be interested in moving bluetooth.rs into a separate crate? It seems very useful as a library. I'm happy to send a PR.)

**Comment by qwandor** [comment]
Hi! I'm glad you've found it useful. The bluetooth.rs module is still a bit of a mess at the moment, and quite incomplete, so I'm not sure it's really in a state to be a separate crate yet, as that might make people think it's actually usable. The better approach long-term might be to move to something like [btleplug](https://crates.io/crates/btleplug) once it's more complete and stable. PRs are welcome though.

As for the WriteValue error, I've seen similar things the first time I try reading a characteristic after connecting to a device. I'm not entirely sure why; maybe it takes BlueZ a bit of time to expose everything over D-Bus. Does it work if you retry? You can try patching in the [retry branch](https://github.com/alsuren/mijia-homie/compare/retry) to see if it helps at all.

**Comment by ruediger** [comment]
Thanks for the quick response! I've tried btleplug as well, but it either simply crashes or can't find devices. I've also tried https://crates.io/crates/bluez but unfortunately it's not very complete and although I can find devices, it lacks GATT functionality.

I'll try the the retry branch and report back.

**Comment by alsuren** [comment]
Hi @ruediger . I noticed that you have a signature mismatch (aya{sv} vs a{sv}). Can you re-run introspect.sh and send me a diff? I don't know what aya{sv} means in dbus signature speak, so I would need to look into it more.

**Comment by ruediger** [comment]
Diff: https://gist.github.com/ruediger/f3c3fbc3604460f845efeb2e1b0ef76b But there is no diff for WriteValue.

aya{sv} means I believe array (a) bytes (y) dict_entry (a{) string (s) -> variant (v) (}). I don't think there is a mismatch, the XML was just weirdly formatted:

```
<method name="WriteValue">
<arg name="value" type="ay" direction="in"/>
<arg name="options" type="a{sv}" direction="in"/>
</method>
```

Which is ay and a{sv}. But just in case it matters I'm using Ubuntu 20.04 which comes with

```
$ bluetoothd -v
5.53
```

I've tried with the retry code and tried adding a longer delay. But I still get the same error. I've tried calling the method in QDBusViewer and I'm also getting an error. So I suspect it's not bluez-generated fault, but some issue I have with Bluez. Unless you have any suggestions or ideas, I think we can close this issue and I'll ask the Bluez folks for help.

Thank you for your help!
